spec/unit/lib/sequel/extensions/connection_metrics_spec.rb

```diff
@@ -0,0 +1,151 @@
+require 'spec_helper'
+
+RSpec.describe Sequel::ConnectionMetrics do
+  let(:db_config) { DbConfig.new }
+  # each test will have their own db connection pool. This helps to isolate the test from each other.
+  let(:db) { VCAP::CloudController::DB.connect(db_config.config, db_config.db_logger) }
+  let(:logs) { StringIO.new }
+  let(:logger) { Logger.new(logs) }
+  let(:prometheus_updater) { spy(VCAP::CloudController::Metrics::PrometheusUpdater) }
+  let(:thread) { double('Thread') }
+
+  before do
+    db.loggers << logger
+    db.sql_log_level = :info
+    db.extension(:connection_metrics)
+
+    allow(thread).to receive(:alive?).and_return(true)
+
+    db.pool.instance_variable_set(:@prometheus_updater, prometheus_updater)
+
+    ENV['PROCESS_TYPE'] = 'puma_worker'
+  end
+
+  after do
+    db.disconnect
+  end
+
+  it 'initializes the prometheus_updater and connection_info' do
+    expect(db.pool.instance_variable_get(:@prometheus_updater)).to eq(prometheus_updater)
+    expect(db.pool.instance_variable_get(:@connection_info)).to be_a(Hash)
+  end
+
+  describe 'acquire' do
+    context 'when acquire returns a connection' do
+      it 'increments the acquired DB connections metric' do
+        expect(prometheus_updater).to receive(:increment_gauge_metric).with(:cc_acquired_db_connections_total, labels: { process_type: 'puma_worker' })
+
+        db.pool.send(:acquire, thread)
+      end
+
+      it 'stores the timestamp when the connection was acquired' do
+        db.pool.send(:acquire, thread)
+        expect(db.pool.instance_variable_get(:@connection_info)[thread]).to have_key(:acquired_at)
+      end
+    end
+
+    context 'when the pool is exhausted and throws a PoolTimeout' do
+      before do
+        # actually acquire throws the exception, but when mocking like this the extension would also be mocked away
+        allow(db.pool).to receive(:assign_connection).and_raise(Sequel::PoolTimeout)
+      end
+
+      it 'increments the connection pool timeouts metric and raises' do
+        expect(prometheus_updater).to receive(:increment_gauge_metric).with(:cc_db_connection_pool_timeouts_total, labels: { process_type: 'puma_worker' })
+
+        expect { db.pool.send(:acquire, thread) }.to raise_error(Sequel::PoolTimeout)
+      end
+
+      it 'emits the time the thread waited for the connection' do
+        db.pool.instance_variable_set(:@connection_info, db.pool.instance_variable_get(:@connection_info).merge!({ thread => { waiting_since: Time.now - 60 } }))
+        expect(prometheus_updater).to receive(:update_histogram_metric).with(:cc_db_connection_wait_duration_seconds, any_args)
```

Comment:
Maybe you can test that it's a duration (i.e. `an_instance_of(ActiveSupport::Duration)`).

Reply:
Good idea! I replaced all `any_args` and `anything` with more precise evaluations.